cmd/osbuild-upload-azure/main.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"net/url"
+	"os"
+	"path"
+
+	"github.com/Azure/azure-storage-blob-go/azblob"
+)
+
+func handleErrors(err error) {
+	if err != nil {
+		if serr, ok := err.(azblob.StorageError); ok { // This error is a Service-specific
+			switch serr.ServiceCode() { // Compare serviceCode to ServiceCodeXxx constants
+			case azblob.ServiceCodeContainerAlreadyExists:
+				// This error is not fatal
+				fmt.Println("Received 409. Container already exists")
+				return
+			}
+		}
+		// All other error causes the program to exit
+		fmt.Println(err)
+		os.Exit(1)
+	}
+}
+
+type azureCredentials struct {
+	storageAccount   string
+	storageAccessKey string
+}
+
+type azureImageMetadata struct {
+	containerName string
+	imageName     string
+}
+
+func azureUploadImage(credentials azureCredentials, metadata azureImageMetadata, fileName string, threads int) {
+	// Create a default request pipeline using your storage account name and account key.
+	credential, err := azblob.NewSharedKeyCredential(credentials.storageAccount, credentials.storageAccessKey)
+	handleErrors(err)
+	p := azblob.NewPipeline(credential, azblob.PipelineOptions{})
+
+	// get storage account blob service URL endpoint.
+	URL, _ := url.Parse(fmt.Sprintf("https://%s.blob.core.windows.net/%s", credentials.storageAccount, metadata.containerName))
+
+	// Create a ContainerURL object that wraps the container URL and a request
+	// pipeline to make requests.
+	containerURL := azblob.NewContainerURL(*URL, p)
+
+	// Create the container, use a never-expiring context
+	ctx := context.Background()
+	_, err = containerURL.Create(ctx, azblob.Metadata{}, azblob.PublicAccessNone)
+	handleErrors(err)
+
+	// Open the image file for reading
+	imageFile, err := os.Open(fileName)
+	handleErrors(err)
+
+	// Stat image to get the file size
+	stat, err := imageFile.Stat()
+	handleErrors(err)
+
+	// Create page blob URL. Page blob is required for VM images
+	blobURL := containerURL.NewPageBlobURL(metadata.imageName)
+	_, err = blobURL.Create(ctx, stat.Size(), 0, azblob.BlobHTTPHeaders{}, azblob.Metadata{}, azblob.BlobAccessConditions{})
+	handleErrors(err)
+
+	// Create control variables
+	// This channel simulates behavior of a semaphore and bounds the number of parallel threads
+	var semaphore = make(chan int, threads)
+	// The page size on Azure is 512, but it was too slow to upload page by page, this is the number
+	// of pages uploaded in single call to "blobURL.UploadPages"
+	var magic int64 = 512
```

Comment:
Why not use the max number of bytes (`PageBlobMaxUploadPagesBytes` (4MB)) rather than talk about pages at all?